app/lib/katello/validators/environment_docker_repositories_validator.rb

```diff
@@ -0,0 +1,78 @@
+module Katello
+  module Validators
+    class EnvironmentDockerRepositoriesValidator < ActiveModel::Validator
+      def validate(environment)
+        return true if environment.registry_name_pattern.empty?
+
+        unless ContainerImageNameValidator.validate_name(Katello::Repository.safe_render_container_name(test_repository, environment.registry_name_pattern))
+          environment.errors.add(:registry_name_pattern, N_("Registry name pattern will result in invalid container image name of member repositories"))
+          return false
+        end
+
+        error_messages = EnvironmentDockerRepositoriesValidator.validate_repositories(environment, environment.registry_name_pattern, environment.repositories.where(content_type: Katello::Repository::DOCKER_TYPE))
+        return true if error_messages.empty?
+
+        error_messages.each do |message|
+          environment.errors.add(:registry_name_pattern, message)
+        end
+        false
+      end
+
+      def self.validate_repositories(environment, registry_name_pattern, repositories)
+        error_messages = []
+        names = []
+        repositories.each do |repository|
+          name = Katello::Repository.safe_render_container_name(repository, registry_name_pattern)
+
+          unless ContainerImageNameValidator.validate_name(name)
+            error_messages << N_("Registry name pattern results in invalid container image name of member repository '%{name}'") % {name: repository.name}
+            return error_messages
+          end
+          names << name
+        end
+
+        if names.length != names.uniq.length
+          error_messages << N_("Registry name pattern results in duplicate container image names")
+          return error_messages
+        end
+
+        Katello::Repository.where(content_type: Katello::Repository::DOCKER_TYPE).each do |repository|
+          if repository.environment != environment && names.include?(repository.container_repository_name)
+            error_messages << N_("Registry name pattern conflicts with an existing repository '%{name}'") % {name: repository.name}
+            return error_messages
+          end
+        end
+
+        []
+      end
+
+      def test_repository
+        Katello::Validators::EnvironmentDockerRepositoriesValidator::RepositoryOpenStruct.new(
+            name: "bad name!", label: "good_label", docker_upstream_name: "image/name", url: "https://registry.example.com",
+            organization: SafeOpenStruct.new(name: "bad name!", label: "good_label"),
+            product: SafeOpenStruct.new(name: "bad name!", label: "good_label"),
+            environment: SafeOpenStruct.new(name: "bad name!", label: "good_label"),
+            content_view_version: OpenStruct.new(content_view: ContentViewOpenStruct.new(name: "bad name!", label: "good_label", version: "1.2"))
+        )
+      end
```

Comment:
was this supposed to be here?

Reply:
Yes, I use this as a fake repo to test the env pattern against as an initial proof. This way things like using unjailed methods, etc. are caught even if the env doesn't have any repos yet.

Reply:
shouldn't this be in the test code though?

Reply:
No, it's used in the validator directly. https://github.com/Katello/katello/pull/7262/files#diff-c334106616785d41f54253e00d15a764R7

I can rename it if you'd prefer. jail_repository or sample_repository or... ?